Adding optional address reference to ERC20Transfer.

paulo-bc opened the pull request:
## Description

Adding optional address reference to ERC20Transfer.

matsjj commented:
@paulo-bc I did a decode using this website

https://flightwallet.github.io/decode-eth-tx/

of your test fixture and found that the function signature is not correct. Some more digging I found that this library actually computes the EVM function signature on the fly based on the arguments that you pass in

https://github.com/oliveratkinson-bc/wallet-core/pull/2/files#diff-ee5abff8d6706ab34d606ef7cfd4679e6da47d2bbb909fd9bf607981c6ed365aR95

This call here effectively takes the function name and the input argument types, hashes them, and takes the first 4 bytes to be the function signature, i.e.

https://piyopiyo.medium.com/how-to-get-ethereum-encoded-function-signatures-1449e171c840

By passing a third argument to the function here
https://github.com/oliveratkinson-bc/wallet-core/pull/2/files#diff-ee5abff8d6706ab34d606ef7cfd4679e6da47d2bbb909fd9bf607981c6ed365aR91
it results in a different hash (`dbba0f01` instead of `a9059cbb`). I believe most smart-contract implementations would just completely ignore this transaction. 

It seems like this function here takes the arguments and parses them into a function signature

https://github.com/oliveratkinson-bc/wallet-core/blob/30bcd54d7f3bdd9d074beef4bd0a31d00c6254e1/src/Ethereum/ABI/Function.cpp#L16

https://github.com/oliveratkinson-bc/wallet-core/blob/30bcd54d7f3bdd9d074beef4bd0a31d00c6254e1/src/Ethereum/ABI/Function.h#L57

https://github.com/oliveratkinson-bc/wallet-core/blob/30bcd54d7f3bdd9d074beef4bd0a31d00c6254e1/src/Ethereum/ABI/Parameters.cpp#L57


In particular, this is exactly what we want to avoid. We make use of the fact that the EVM ignores *extra* data beyond what is specified in the function signature. We need to find a way to make this library accept extra parameters without encoding them. For example we can probably add a boolean here

https://github.com/oliveratkinson-bc/wallet-core/blob/30bcd54d7f3bdd9d074beef4bd0a31d00c6254e1/src/Ethereum/ABI/Parameters.h#L38

Of the form `shouldEncodeInFunctionSignature()` or similar, that is checked here
https://github.com/oliveratkinson-bc/wallet-core/blob/30bcd54d7f3bdd9d074beef4bd0a31d00c6254e1/src/Ethereum/ABI/Parameters.cpp#L62

paulo-bc commented:
> @paulo-bc I did a decode using this website
> 
> https://flightwallet.github.io/decode-eth-tx/
> 
> of your test fixture and found that the function signature is not correct. Some more digging I found that this library actually computes the EVM function signature on the fly based on the arguments that you pass in
> 
> https://github.com/oliveratkinson-bc/wallet-core/pull/2/files#diff-ee5abff8d6706ab34d606ef7cfd4679e6da47d2bbb909fd9bf607981c6ed365aR95
> 
> This call here effectively takes the function name and the input argument types, hashes them, and takes the first 4 bytes to be the function signature, i.e.
> 
> https://piyopiyo.medium.com/how-to-get-ethereum-encoded-function-signatures-1449e171c840
> 
> By passing a third argument to the function here https://github.com/oliveratkinson-bc/wallet-core/pull/2/files#diff-ee5abff8d6706ab34d606ef7cfd4679e6da47d2bbb909fd9bf607981c6ed365aR91 it results in a different hash (`dbba0f01` instead of `a9059cbb`). I believe most smart-contract implementations would just completely ignore this transaction.
> 
> It seems like this function here takes the arguments and parses them into a function signature
> 
> https://github.com/oliveratkinson-bc/wallet-core/blob/30bcd54d7f3bdd9d074beef4bd0a31d00c6254e1/src/Ethereum/ABI/Function.cpp#L16
> 
> https://github.com/oliveratkinson-bc/wallet-core/blob/30bcd54d7f3bdd9d074beef4bd0a31d00c6254e1/src/Ethereum/ABI/Function.h#L57
> 
> https://github.com/oliveratkinson-bc/wallet-core/blob/30bcd54d7f3bdd9d074beef4bd0a31d00c6254e1/src/Ethereum/ABI/Parameters.cpp#L57
> 
> In particular, this is exactly what we want to avoid. We make use of the fact that the EVM ignores _extra_ data beyond what is specified in the function signature. We need to find a way to make this library accept extra parameters without encoding them. For example we can probably add a boolean here
> 
> https://github.com/oliveratkinson-bc/wallet-core/blob/30bcd54d7f3bdd9d074beef4bd0a31d00c6254e1/src/Ethereum/ABI/Parameters.h#L38
> 
> Of the form `shouldEncodeInFunctionSignature()` or similar, that is checked here
> 
> https://github.com/oliveratkinson-bc/wallet-core/blob/30bcd54d7f3bdd9d074beef4bd0a31d00c6254e1/src/Ethereum/ABI/Parameters.cpp#L62

@matsjj thank you for the in depth review, I am getting back to this today and will get this all done 🙇

paulo-bc commented:
@matsjj Hey Mats, PR ready for review again.
I added a `shouldSkipEncodeInFunctionSignature` parameter to `ParamBase` (default false), added test covering that `getType()` behaves correctly with this change, and fixed the test you referenced that now has the fn hash `a9059cbb`.

oliveratkinson-bc commented on src/Ethereum/ABI/Parameters.cpp:
```diff
@@ -59,7 +59,11 @@ std::string ParamSet::getType() const {
     for (auto p : _params) {
-        if (cnt++ > 0) {
+        bool isFirst = cnt++ > 0;
+        if (p->shouldSkipEncodeInFunctionSignature) {
```
We should look to submit this P/R upstream to get feedback from the trustwallet maintainers - I do not understand this enough to know if altering the Ethereum ABI is safe to do

matsjj commented on src/interface/TWEthereumAbiFunction.cpp:
```diff
@@ -74,6 +74,16 @@ int TWEthereumAbiFunctionAddParamUInt64(struct TWEthereumAbiFunction *_Nonnull f
     return idx;
 }
 
+int TWEthereumAbiFunctionAddParamUInt64SkipEncode(struct TWEthereumAbiFunction *_Nonnull func_in, uint64_t val, bool isOutput) {
```
Did you only add this function in here to simplify testing? In that case I think we should try to push it into the tests.

paulo-bc replied:
This is used only in tests, but it could be used by any consumer library. Same applies to all other pre-existing `TWEthereumAbiFunctionAddParam...` methods.

matsjj commented on swift/Tests/Blockchains/EthereumTests.swift:
```diff
@@ -64,6 +64,29 @@ class EthereumTests: XCTestCase {
+        let output: EthereumSigningOutput = AnySigner.sign(input: input, coin: .ethereum)
+
+        XCTAssertEqual(output.encoded.hexString, "f8ca808509c7652400830130b9946b175474e89094c44da98b954eedeac495271d0f80b864dbba0f010000000000000000000000005322b34c88ed0691971bf52a7047448f0f4efc840000000000000000000000000000000000000000000000001bc16d674ec80000000000000000000000000000353535353535353535353535353535353535353526a06df79d65a3c2e7564a5d569ccd9a07f868b181b67e2b01d84eab0b5828e7b006a072355 ...
+        XCTAssertEqual(output.data.hexString, "dbba0f010000000000000000000000005322b34c88ed0691971bf52a7047448f0f4efc840000000000000000000000000000000000000000000000001bc16d674ec800000000000000000000000000003535353535353535353535353535353535353535")
```
You probably want to adjust this, as right now this test should not pass anymore?

paulo-bc replied:
Will build the swift framework and check this out.

matsjj commented on tests/Ethereum/TWEthereumAbiTests.cpp:
```diff
@@ -51,6 +51,29 @@ TEST(TWEthereumAbi, FuncCreate1) {
     TWEthereumAbiFunctionDelete(func);
 }
 
+TEST(TWEthereumAbi, FuncCreateSkip) {
```
I don't quite understand what this is testing, can you please add a comment? 🙏

paulo-bc replied:
Done.

paulo-bc commented on src/Ethereum/ABI/Parameters.cpp:
```diff
@@ -59,7 +59,11 @@ std::string ParamSet::getType() const {
     int cnt = 0;
     for (auto p : _params) {
-        if (cnt++ > 0) {
+        bool isFirst = cnt++ > 0;
```
Hey @matsjj , I can't undo this as we need a small change here because when `p->shouldSkipEncodeInFunctionSignature` we should still increment the `cnt` but we should not add a comma `,`.
Maybe you have a better formatting in mind?
